### PR TITLE
Make location queryset order by tree and name

### PR DESCRIPTION
**Pull request by esoergel**
@sravfeyn @mkangia @proteusvacuum
I just figured out that this was possible - it tells MPTT to factor in the "name" field when setting the "lft" property, which means that with the [default ordering](https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/locations/models.py#L316-L316), we'll get something sensible and performant.

What this means is that the default ordering is probably what we should use in most cases, like for displaying in filters and such.  I'm pinging you three just to spread awareness of this, since it's not very obvious that this works at all, let alone how it works.

https://django-mptt.github.io/django-mptt/mptt.models.html#mptt.models.MPTTOptions

@dannyroberts buddy

### CODE REVIEW

**Comment by dimagimon** [comment]
There are 13 errors:

* corehq/apps/locations/tests/test_location_queries.py, line 145 - F841 local variable 'dorcester' is assigned to but never used
* corehq/apps/locations/tests/test_location_queries.py, line 146 - F841 local variable 'boston' is assigned to but never used
* corehq/apps/locations/tests/test_location_queries.py, line 148 - F841 local variable 'somerville' is assigned to but never used
* corehq/apps/locations/tests/test_location_queries.py, line 149 - F841 local variable 'cambridge' is assigned to but never used
* corehq/apps/locations/tests/test_location_queries.py, line 150 - F841 local variable 'arlington' is assigned to but never used
* corehq/apps/locations/tests/test_location_queries.py, line 157 - E131 continuation line unaligned for hanging indent
* corehq/apps/locations/tests/test_location_queries.py, line 158 - E131 continuation line unaligned for hanging indent
* corehq/apps/locations/tests/test_location_queries.py, line 173 - F841 local variable 'arlington' is assigned to but never used
* corehq/apps/locations/tests/test_location_queries.py, line 174 - F841 local variable 'boston' is assigned to but never used
* corehq/apps/locations/tests/test_location_queries.py, line 175 - F841 local variable 'dorcester' is assigned to but never used
* corehq/apps/locations/tests/test_location_queries.py, line 176 - F841 local variable 'somerville' is assigned to but never used
* corehq/apps/locations/tests/test_location_queries.py, line 187 - E131 continuation line unaligned for hanging indent
* corehq/apps/locations/tests/test_location_queries.py, line 188 - E131 continuation line unaligned for hanging indent

**Comment by sravfeyn** [comment]
Looks good. Build fail looks real

File: corehq/apps/locations/tests/test_location_queries.py [hunk 1]
```diff
@@ -126,3 +126,70 @@ def test_filter_path_by_user_input(self):
...
+            # I know the indentation is weird - it's here to illustrate WHY this is the correct ordering
+            [
+                "Mass",
+                    "Middlesex",
+                        "Arlington",
+                        "Cambridge",
+                        "Somerville",
+                    "Suffolk",
+                        "Boston",
+                        "Dorcester",
```
Comment (proteusvacuum) [inline corehq/apps/locations/tests/test_location_queries.py]: nit: Dorchester

**Comment by proteusvacuum** [comment]
cool!

**Comment by esoergel** [comment]
Ouch, this build failure is really tough.  Setting this meta property causes MPTT to use different code paths, and I think we're running against an unhandled edge-case.  `get_next_sibling()` is returning `None` [here](https://github.com/django-mptt/django-mptt/blob/master/mptt/managers.py#L942).  That's expected behavior for that method (it's mentioned in the docstring), but the call doesn't handle the null case, so it fails hard with a `'NoneType' has no attribute...` error.

I'm gonna give up trying to figure this out for now, but feel free to pick up the torch if anyone wants a crack at it.  Easily reproducible with:
```
./manage.py test corehq.apps.locations.tests.test_bulk_management:TestBulkManagement.test_edit_names
```

File: corehq/apps/locations/models.py [hunk 2]
```diff
@@ -592,6 +592,9 @@ class Meta:
         app_label = 'locations'
         unique_together = ('domain', 'site_code',)
 
+    class MPTTMeta:
+        order_insertion_by = ['name']
```
Comment (esoergel) [inline corehq/apps/locations/models.py]: `django-mptt` essentially precomputes the order and saves that to the database.  It's done in python before saving, not in SQL during queries (AFAIK).

**Comment by dimagimon** [comment]
There are 13 errors:

* corehq/apps/locations/tests/test_location_queries.py, line 145 - F841 local variable 'dorchester' is assigned to but never used
* corehq/apps/locations/tests/test_location_queries.py, line 146 - F841 local variable 'boston' is assigned to but never used
* corehq/apps/locations/tests/test_location_queries.py, line 148 - F841 local variable 'somerville' is assigned to but never used
* corehq/apps/locations/tests/test_location_queries.py, line 149 - F841 local variable 'cambridge' is assigned to but never used
* corehq/apps/locations/tests/test_location_queries.py, line 150 - F841 local variable 'arlington' is assigned to but never used
* corehq/apps/locations/tests/test_location_queries.py, line 157 - E131 continuation line unaligned for hanging indent
* corehq/apps/locations/tests/test_location_queries.py, line 158 - E131 continuation line unaligned for hanging indent
* corehq/apps/locations/tests/test_location_queries.py, line 173 - F841 local variable 'arlington' is assigned to but never used
* corehq/apps/locations/tests/test_location_queries.py, line 174 - F841 local variable 'boston' is assigned to but never used
* corehq/apps/locations/tests/test_location_queries.py, line 175 - F841 local variable 'dorchester' is assigned to but never used
* corehq/apps/locations/tests/test_location_queries.py, line 176 - F841 local variable 'somerville' is assigned to but never used
* corehq/apps/locations/tests/test_location_queries.py, line 187 - E131 continuation line unaligned for hanging indent
* corehq/apps/locations/tests/test_location_queries.py, line 188 - E131 continuation line unaligned for hanging indent

File: corehq/apps/locations/tests/test_location_queries.py (hunk 1 above)
Comment (esoergel) [inline corehq/apps/locations/tests/test_location_queries.py]: :astonished:

**Comment by millerdev** [comment]
@snopoke @esoergel The plan is to have CTE (descendants) queries order hierarchically

https://github.com/dimagi/commcare-hq/blob/e8642ba1ae5ad55044d26af1ecaf3640f7fad7b6/corehq/apps/locations/adjacencylist.py#L164

by name

https://github.com/dimagi/commcare-hq/blob/e8642ba1ae5ad55044d26af1ecaf3640f7fad7b6/corehq/apps/locations/adjacencylist.py#L226

once MPTT is removed.

Ancestor query ordering will not change, although it is implemented in a new way:

https://github.com/dimagi/commcare-hq/blob/e8642ba1ae5ad55044d26af1ecaf3640f7fad7b6/corehq/apps/locations/adjacencylist.py#L82

There's a gotcha with this one though: `get_queryset_ancestors` does not have a nice ordering because it groups each depth value together when there are multiple trees in the queryset. I'm not sure that matters anywhere though.

File: corehq/apps/locations/models.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (millerdev) [reply corehq/apps/locations/models.py]: I'm assuming we would need to update rows in the database to make this effective for existing locations. Is that correct, and is there a plan to do that?

**Comment by esoergel** [comment]
@snopoke @millerdev I don't trust the MPTT ordering, so I'd say this PR will be superceded by the adjancency list work.  This is an old PR (December 2016) that I left open on the hopes that MPTT would fix [the bug I identified](https://github.com/django-mptt/django-mptt/pull/530), making it mergeable.  Since we're dropping MPTT, the tests here may be useful, but the insertion order change isn't.

@millerdev I can't think of any places where the ordering of `get_queryset_ancestors` might matter, as you suggested.

File: corehq/apps/locations/models.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (esoergel) [reply corehq/apps/locations/models.py]: I don't plan to merge this change as-is - instead, I'll wait until we drop MPTT and maybe use these tests with the adjancency list work if they seem useful.

**Pull request closed without merging**